client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/iteration/RemoteCloseableIterator.java

```diff
@@ -76,30 +75,30 @@ public boolean hasNext() {
    }
 
    @Override
-   public Entry<Object, Object> next() {
+   public Entry<K, V> next() {
       if (!hasNext()) throw new NoSuchElementException();
       return nextElements.remove();
    }
 
    private void fetch() {
       try {
-         IterationNextOperation iterationNextOperation = operationsFactory.newIterationNextOperation(iterationId, transport);
+         IterationNextOperation<K, V> iterationNextOperation = operationsFactory.newIterationNextOperation(iterationId, transport);
 
          while (nextElements.isEmpty() && !endOfIteration) {
-            IterationNextResponse iterationNextResponse = iterationNextOperation.execute();
+            IterationNextResponse<K, V> iterationNextResponse = iterationNextOperation.execute();
             short status = iterationNextResponse.getStatus();
-            if (status == INVALID_ITERATION) {
+            if (HotRodConstants.isInvalidIteration(status)) {
                throw log.errorRetrievingNext(iterationId);
             }
-            Entry<byte[], byte[]>[] entries = iterationNextResponse.getEntries();
+            Entry<K, V>[] entries = iterationNextResponse.getEntries();
 
             if (entries.length == 0) {
                endOfIteration = true;
                break;
             }
-            for (Entry<byte[], byte[]> entry : entries) {
-               if (segmentKeyTracker.track(entry.getKey())) {
-                  nextElements.add(new SimpleEntry<>(unmarshall(entry.getKey()), unmarshall(entry.getValue())));
+            for (Entry<K, V> entry : entries) {
+               if (segmentKeyTracker.track(obj2bytes(marshaller, entry.getKey(), true, 100, 100))) {
```

Comment:
Can this extra marshalling be avoided?
On master, it takes raw entries from ```iterationNextOperation```, uses the ```segmentKeyTracker``` to track raw keys, and finally unmarshalls K,V to serve to the user. Now it is always unmarshalling K and V plus marshalling again the key. :/